### PR TITLE
Portable comparisons (for sorting)

### PR DESCRIPTION
**Pull request by QuLogic**
There are a few mixed comparisons (`None` with `int`, `str` with `int`, etc.) that do not work directly with Python 3. So add an extra element to the sorting key so that those comparisons do not occur (and work similar to Python 2).

### CODE REVIEW

File: lib/iris/coords.py [hunk 1]
```diff
@@ -64,6 +64,23 @@ def name(self, default='unknown'):
...
+    def __lt__(self, other):
+        if not isinstance(other, CoordDefn):
+            return NotImplemented
+
+        # Emulate Python 2 behaviour with None
+        key = (self.standard_name is not None, self.standard_name or '',
+               self.long_name is not None, self.long_name or '',
+               self.var_name is not None, self.var_name or '',
+               self.units is not None, self.units or '',
+               self.coord_system is not None, self.coord_system or '')
```
Comment (QuLogic) [inline lib/iris/coords.py]: Does it even make sense to use the units or coordinate system in the comparison? I'm not even sure they're comparable. Should I leave them out?

Comment (QuLogic) [inline lib/iris/coords.py]: Also, thinking about it a bit, those first tests should ensure never comparing `None` with `str`, so maybe I can leave out those `or ''`?

**Comment by QuLogic** [comment]
Oh, I forgot I added some usage of `six`, which means this PR will conflict with #1700. Please merge it soon.

File: lib/iris/_merge.py [hunk 2]
```diff
@@ -1300,7 +1300,11 @@ def _define_space(self, space, positions, indexes, function_matrix):
         def axis_and_name(name):
             axis_dict = {'T': 1, 'Z': 2, 'Y': 3, 'X': 4}
             axis_index = axis_dict.get(self._guess_axis(name), 0)
-            return (axis_index, name)
+            # The middle element ensures sorting is the same as Python 2.
+            if isinstance(name, six.integer_types):
+                return (axis_index, 0, str(name))
+            else:
+                return (axis_index, 1, name)
```
Comment (rhattersley) [inline lib/iris/_merge.py]: Could this `if` be replaced with `return (axis_index, isinstance(name, int), str(name))`?

File: lib/iris/coords.py (hunk 1 above)
Comment (rhattersley) [inline lib/iris/coords.py]: > Does it even make sense to use the units or coordinate system in the comparison?

Yes - so we can have consistent (albeit arbitrary) ordering for coordinates which are not equal.



> maybe I can leave out those `or ''`?

:+1:

File: lib/iris/coords.py [hunk 3]
```diff
@@ -64,6 +64,23 @@ def name(self, default='unknown'):
...
+        okey = (other.standard_name is not None, other.standard_name or '',
+                other.long_name is not None, other.long_name or '',
+                other.var_name is not None, other.var_name or '',
+                other.units is not None, other.units or '',
+                other.coord_system is not None, other.coord_system or '')
```
Comment (rhattersley) [inline lib/iris/coords.py]: Perhaps refactor into an inner function?

File: lib/iris/cube.py [hunk 4]
```diff
@@ -452,9 +452,16 @@ def merge(self, unique=True):
                 proto_cube = iris._merge.ProtoCube(cube)
                 proto_cubes.append(proto_cube)
 
+        # Emulate Python 2 behaviour.
+        def _none_sort(item):
+            if item is None:
+                return (0, '')
+            else:
+                return (1, item)
```
Comment (rhattersley) [inline lib/iris/cube.py]: `return (item is not None, item)`?

File: lib/iris/tests/unit/fileformats/grib/load_rules/test_convert.py [hunk 5]
```diff
@@ -95,7 +95,8 @@ def assert_bounded_message(self, **kwargs):
                       'edition': 1, '_forecastTime': 15,
                       '_forecastTimeUnit': 'hours',
                       'phenomenon_bounds': lambda u: (80, 120),
-                      '_phenomenonDateTime': -1}
+                      '_phenomenonDateTime': -1,
+                      'table2Version': 9999}
```
Comment (rhattersley) [inline lib/iris/tests/unit/fileformats/grib/load_rules/test_convert.py]: Is this change supposed to be here?

Comment (QuLogic) [inline lib/iris/tests/unit/fileformats/grib/load_rules/test_convert.py]: Yes, somewhere in the GRIB loader is `if grib.table2Version < 128:` which doesn't work with a `Mock` object (not exactly sorting, but related to comparison operators.)

File: lib/iris/_merge.py (hunk 2 above)
Comment (QuLogic) [inline lib/iris/_merge.py]: Yes, I think so; probably can leave out the `str` to for the same reason as above.

File: lib/iris/tests/unit/fileformats/grib/load_rules/test_convert.py (hunk 5 above)
Comment (rhattersley) [inline lib/iris/tests/unit/fileformats/grib/load_rules/test_convert.py]: OK - thanks for clarifying. :+1:

**Comment by QuLogic** [comment]
I added one more sorting fix, 184e165773b165c0235c24efa100189f902d9d7e.

File: lib/iris/io/format_picker.py [hunk 6]
```diff
@@ -214,7 +214,7 @@ def __lt__(self, other):
         if not isinstance(other, FormatSpecification):
             return NotImplemented
 
-        return (-self.priority, hash(self)) < (-other.priority, hash(other))
+        return (-self.priority, self.name) < (-other.priority, other.name)
```
Comment (QuLogic) [inline lib/iris/io/format_picker.py]: Hmm, now that I think about it, this one may only work if names are unique? (because `__eq__` still uses `hash`.)

Comment (pelson) [inline lib/iris/io/format_picker.py]: Agreed. This is not a unique hash.

**Comment by QuLogic** [comment]
Maybe that will work better.

**Comment by QuLogic** [comment]
So I've found two more comparison issues:
```
======================================================================
ERROR: test_multiple_unordered_lbprocs (iris.tests.unit.fileformats.pp_rules.test__all_other_rules.TestCellMethods)
----------------------------------------------------------------------
Traceback (most recent call last):
  File "/home/elliott/code/iris/testdir/conda/envs/env34/lib/python3.4/site-packages/Iris-1.9.0.dev0-py3.4.egg/iris/tests/unit/fileformats/pp_rules/test__all_other_rules.py", line 125, in test_multiple_unordered_lbprocs
    res = _all_other_rules(field)[CELL_METHODS_INDEX]
  File "/home/elliott/code/iris/testdir/conda/envs/env34/lib/python3.4/site-packages/Iris-1.9.0.dev0-py3.4.egg/iris/fileformats/pp_rules.py", line 881, in _all_other_rules
    coord_system=f.coord_system()),
  File "/home/elliott/code/iris/testdir/conda/envs/env34/lib/python3.4/site-packages/Iris-1.9.0.dev0-py3.4.egg/iris/coords.py", line 1346, in from_regular
    if not is_regular(coord) and count > 1:
TypeError: unorderable types: MagicMock() > int()

======================================================================
ERROR: test_multiple_unordered_rotated_lbprocs (iris.tests.unit.fileformats.pp_rules.test__all_other_rules.TestCellMethods)
----------------------------------------------------------------------
Traceback (most recent call last):
  File "/home/elliott/code/iris/testdir/conda/envs/env34/lib/python3.4/site-packages/Iris-1.9.0.dev0-py3.4.egg/iris/tests/unit/fileformats/pp_rules/test__all_other_rules.py", line 136, in test_multiple_unordered_rotated_lbprocs
    res = _all_other_rules(field)[CELL_METHODS_INDEX]
  File "/home/elliott/code/iris/testdir/conda/envs/env34/lib/python3.4/site-packages/Iris-1.9.0.dev0-py3.4.egg/iris/fileformats/pp_rules.py", line 881, in _all_other_rules
    coord_system=f.coord_system()),
  File "/home/elliott/code/iris/testdir/conda/envs/env34/lib/python3.4/site-packages/Iris-1.9.0.dev0-py3.4.egg/iris/coords.py", line 1346, in from_regular
    if not is_regular(coord) and count > 1:
TypeError: unorderable types: MagicMock() > int()
```

I'm not sure what the expected value should be for these. In those two tests, some fields of the `pp` mocked data are unset. @marqh Can you shed some light on these since you wrote the tests?

**Comment by marqh** [comment]
the tests that are failing are failing in the iris.coords

the error is due to the fact that a mock object is being evaluated compared to an integer.  this is because the tests provide a mocked version of the _x_coord_+name and _y_coord_name in the mock field

most mock tests only use the parts of the PP field they require

@QuLogic please may you be more specific when you say:
> I'm not sure what the expected value should be for these.
which expected values? for which 'these'?

thank you
mark

**Comment by QuLogic** [comment]
The [PP loader calls `DimCoord.from_regular(f.bzx, f.bdx, f.lbnpt, ...`](https://github.com/SciTools/iris/blob/master/lib/iris/fileformats/pp_rules.py#L876-L880), but none of these fields are set in the mocked object. Those are the fields I mean.

**Comment by marqh** [comment]
> The PP loader calls DimCoord.from_regular(f.bzx, f.bdx, f.lbnpt, ..., but none of these fields are set in the mocked object. Those are the fields I mean.

I see.  Those metadata elements are not required for this test, the only important facet for setting the cell_methods for the lbproc codes is whether the grid is rotated or not, so I mocked out the behaviour with the _{x|y}_coord_name 

I'm afraid I don't understand the nature of the test failure. I can see that the `coord_system` is being run on the field, but I don't see why this is now failing

we could make up a set of arbitrary values for the field to enable this test to pass, although this adds complexity to the test

**Comment by ajdawson** [comment]
Perhaps this was happening all along but on Python 2 the comparison just returns `False`? See below.

Python 3:

```python
>>> from mock import MagicMock
>>> m = MagicMock()
>>> m > 1
Traceback (most recent call last):
  File "<stdin>", line 1, in <module>
TypeError: unorderable types: MagicMock() > int()
```

Python 2:

```python
>>> from mock import MagicMock
>>> m = MagicMock()
>>> m > 1
False
```

**Comment by QuLogic** [comment]
Yes, that's correct; Python 2 sorts them arbitrarily. If the actual points don't really matter, how about just setting them to 0?

**Comment by QuLogic** [comment]
> Those metadata elements are not required for this test

I'm not sure that's true. The code from [here](https://github.com/SciTools/iris/blob/master/lib/iris/fileformats/pp_rules.py#L873) onward has several conditions that depend on `bdx` and `bdy`, and are depending on mock's automatic attribute behaviour. But `lbcode` _is_ set to select the right conditions earlier, so I think `bdx` and `bdy` should be set similarly.

File: lib/iris/tests/results/file_load/known_loaders.txt [hunk 7]
```diff
@@ -1,16 +1,16 @@
  * NetCDF OPeNDAP (priority 6)
  * NAME III (priority 5)
- * NetCDF_v4 (priority 5)
- * UM Post Processing file (PP) (priority 5)
  * NetCDF (priority 5)
  * NetCDF 64 bit offset format (priority 5)
+ * NetCDF_v4 (priority 5)
+ * UM Post Processing file (PP) (priority 5)
```
Comment (rhattersley) [inline lib/iris/tests/results/file_load/known_loaders.txt]: We don't really want this list changing order ... that indicates a potential change in behaviour.

Comment (rhattersley) [inline lib/iris/tests/results/file_load/known_loaders.txt]: The order comes from the `FormatSpecification` comparitor:
```python
    def __lt__(self, other):
        if not isinstance(other, FormatSpecification):
            return NotImplemented

        return (-self.priority, hash(self)) < (-other.priority, hash(other))
```
Eww :scream:

Comment (rhattersley) [inline lib/iris/tests/results/file_load/known_loaders.txt]: Perhaps the simplest solution is to tweak the definitions in `lib/iris/fileformats/__init__.py` to use unique numbers that preserve the current order?

**Comment by rhattersley** [comment]
> I think bdx and bdy should be set similarly.

:+1: Yes, that makes sense. Just as we don't really care about `_x_coord_name` when checking cell methods, we don't care about `bdx`, but we need to set both just to get through that routine. (You might need to set `bzx`, `lbnpt`, `bzy` and `lbrow` as well.)

FWIW, I'd set `bdx` and `bdy` to 1. If you need to set `bzx` and `bzy` I'd set them to 0. I'd set `lbnpt` and `lbrow` to 3.

File: lib/iris/tests/results/file_load/known_loaders.txt (hunk 7 above)
Comment (QuLogic) [inline lib/iris/tests/results/file_load/known_loaders.txt]: I assume (perhaps incorrectly) that as long as they are at the same priority, it should not matter which one is ordered first. I don't think the tests have complained so far...

Comment (rhattersley) [inline lib/iris/tests/results/file_load/known_loaders.txt]: Format specifications with the same priority *should* have mutually exclusive file matching criteria, in which case we would be able to get away with it. I'll check...

Comment (rhattersley) [inline lib/iris/tests/results/file_load/known_loaders.txt]: Mostly they are mutually exclusive, but sadly not in the case of the ABF/L formats vs. the other "priority 3" formats. For example, a FieldsFile named "foo.abf" would load OK using the old order, but would try to load as an ABF file using the new order.

Comment (rhattersley) [inline lib/iris/tests/results/file_load/known_loaders.txt]: NB. We really shouldn't be relying on consistent hash results anyway - e.g. http://bugs.python.org/issue13703.

Comment (rhattersley) [inline lib/iris/tests/results/file_load/known_loaders.txt]: Given the minimal impact I suggest we just accept the change as it makes the order more predictable and controllable.

**Comment by QuLogic** [comment]
Sorry for the delay; I needed to set a few of the ones you mentioned plus `x_bounds=None` so it wouldn't get confused setting an inconsistent bounds.
